File: MiniKeePass/CharacterSetsViewController.m
/*
 * Copyright 2011-2012 Jason Rush and John Flanagan. All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "CharacterSetsViewController.h"
#import "AppSettings.h"

#define NUMBER_CHARACTER_SETS    8

@implementation CharacterSetsViewController

- (id)init {
    self = [super initWithStyle:UITableViewStyleGrouped];
    if (self) {
        self.title = NSLocalizedString(@"Character Sets", nil);

        NSInteger charSets = [[AppSettings sharedInstance] pwGenCharSets];
        
        upperCaseSwitchCell = [[SwitchCell alloc] initWithLabel:NSLocalizedString(@"Upper Case", nil)];
        upperCaseSwitchCell.switchControl.on = charSets & CHARACTER_SET_UPPER_CASE;
        
        lowerCaseSwitchCell = [[SwitchCell alloc] initWithLabel:NSLocalizedString(@"Lower Case", nil)];
        lowerCaseSwitchCell.switchControl.on = charSets & CHARACTER_SET_LOWER_CASE;
        
        digitsSwitchCell = [[SwitchCell alloc] initWithLabel:NSLocalizedString(@"Digits", nil)];
        digitsSwitchCell.switchControl.on = charSets & CHARACTER_SET_DIGITS;
        
        minusSwitchCell = [[SwitchCell alloc] initWithLabel:NSLocalizedString(@"Minus", nil)];
        minusSwitchCell.switchControl.on = charSets & CHARACTER_SET_MINUS;
        
        underlineSwitchCell = [[SwitchCell alloc] initWithLabel:NSLocalizedString(@"Underline", nil)];
        underlineSwitchCell.switchControl.on = charSets & CHARACTER_SET_UNDERLINE;
        
        spaceSwitchCell = [[SwitchCell alloc] initWithLabel:NSLocalizedString(@"Space", nil)];
        spaceSwitchCell.switchControl.on = charSets & CHARACTER_SET_SPACE;
        
        specialSwitchCell = [[SwitchCell alloc] initWithLabel:NSLocalizedString(@"Special", nil)];
        specialSwitchCell.switchControl.on = charSets & CHARACTER_SET_SPECIAL;
        
        bracketsSwitchCell = [[SwitchCell alloc] initWithLabel:NSLocalizedString(@"Brackets", nil)];
        bracketsSwitchCell.switchControl.on = charSets & CHARACTER_SET_BRACKETS;
    }
    return self;
}

- (void)viewWillDisappear:(BOOL)animated {
    NSInteger charSets = 0;
    if (upperCaseSwitchCell.switchControl.on) {
        charSets |= CHARACTER_SET_UPPER_CASE;
    }
    if (lowerCaseSwitchCell.switchControl.on) {
        charSets |= CHARACTER_SET_LOWER_CASE;
    }
    if (digitsSwitchCell.switchControl.on) {
        charSets |= CHARACTER_SET_DIGITS;
    }
    if (minusSwitchCell.switchControl.on) {
        charSets |= CHARACTER_SET_MINUS;
    }
    if (underlineSwitchCell.switchControl.on) {
        charSets |= CHARACTER_SET_UNDERLINE;
    }
    if (spaceSwitchCell.switchControl.on) {
        charSets |= CHARACTER_SET_SPACE;
    }
    if (specialSwitchCell.switchControl.on) {
        charSets |= CHARACTER_SET_SPECIAL;
    }
    if (bracketsSwitchCell.switchControl.on) {
        charSets |= CHARACTER_SET_BRACKETS;
    }
    
    [[AppSettings sharedInstance] setPwGenCharSets:charSets];
    
    [super viewWillDisappear:animated];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return NUMBER_CHARACTER_SETS;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    switch (1 << indexPath.row) {
        case CHARACTER_SET_UPPER_CASE:
            return upperCaseSwitchCell;
        case CHARACTER_SET_LOWER_CASE:
            return lowerCaseSwitchCell;
        case CHARACTER_SET_DIGITS:
            return digitsSwitchCell;
        case CHARACTER_SET_MINUS:
            return minusSwitchCell;
        case CHARACTER_SET_UNDERLINE:
            return underlineSwitchCell;
        case CHARACTER_SET_SPACE:
            return spaceSwitchCell;
        case CHARACTER_SET_SPECIAL:
            return specialSwitchCell;
        case CHARACTER_SET_BRACKETS:
            return bracketsSwitchCell;
        default:
            break;
    }
    
    return nil;
}

@end

File: MiniKeePass/Password Generator View/CharacterSetsViewController.swift
/*
 * Copyright 2016 Jason Rush and John Flanagan. All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import UIKit

struct CharacterSet {
    static let UpperCase = 1 << 0
    static let LowerCase = 1 << 1
    static let Digits    = 1 << 2
    static let Minus     = 1 << 3
    static let Underline = 1 << 4
    static let Space     = 1 << 5
    static let Special   = 1 << 6
    static let Brackets  = 1 << 7
    static let Default   = (UpperCase | LowerCase | Digits)
}

class CharacterSetsViewController: UITableViewController {
    @IBOutlet weak var upperCaseSwitch: UISwitch!
    @IBOutlet weak var lowerCaseSwitch: UISwitch!
    @IBOutlet weak var digitsSwitch: UISwitch!
    @IBOutlet weak var minusSwitch: UISwitch!
    @IBOutlet weak var underlineSwitch: UISwitch!
    @IBOutlet weak var spaceSwitch: UISwitch!
    @IBOutlet weak var specialSwitch: UISwitch!
    @IBOutlet weak var bracketsSwitch: UISwitch!

    override func viewDidLoad() {
        super.viewDidLoad()

        let appSettings = AppSettings.sharedInstance() as AppSettings
        let charSets = appSettings.pwGenCharSets()
        
        upperCaseSwitch.isOn = (charSets & CharacterSet.UpperCase) != 0
        lowerCaseSwitch.isOn = (charSets & CharacterSet.LowerCase) != 0
        digitsSwitch.isOn = (charSets & CharacterSet.Digits) != 0
        minusSwitch.isOn = (charSets & CharacterSet.Minus) != 0
        underlineSwitch.isOn = (charSets & CharacterSet.Underline) != 0
        spaceSwitch.isOn = (charSets & CharacterSet.Space) != 0
        specialSwitch.isOn = (charSets & CharacterSet.Special) != 0
        bracketsSwitch.isOn = (charSets & CharacterSet.Brackets) != 0
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        var charSets = 0
        if (upperCaseSwitch.isOn) {
            charSets |= CharacterSet.UpperCase
        }
        if (lowerCaseSwitch.isOn) {
            charSets |= CharacterSet.LowerCase
        }
        if (digitsSwitch.isOn) {
            charSets |= CharacterSet.Digits
        }
        if (minusSwitch.isOn) {
            charSets |= CharacterSet.Minus
        }
        if (underlineSwitch.isOn) {
            charSets |= CharacterSet.Underline
        }
        if (spaceSwitch.isOn) {
            charSets |= CharacterSet.Space
        }
        if (specialSwitch.isOn) {
            charSets |= CharacterSet.Special
        }
        if (bracketsSwitch.isOn) {
            charSets |= CharacterSet.Brackets
        }
        
        let appSettings = AppSettings.sharedInstance() as AppSettings
        appSettings.setPwGenCharSets(charSets)
    }
}
